src/compiler/checker.ts

```diff
@@ -27332,7 +27324,7 @@ namespace ts {
 
         function checkDeclarationInitializer(declaration: HasExpressionInitializer) {
             const initializer = getEffectiveInitializer(declaration)!;
-            const type = getTypeOfExpression(initializer, /*cache*/ true);
+            const type = getQuickTypeOfExpression(initializer) || checkExpressionCached(initializer);

```

Comment:
IIRC, using the `cache` flag here was due to a perf issue, not a correctness one, and the caching was a correctness trade-off - using the per-loop cache is probably better, if possible.

Reply:
No, we need to call `checkExpressionCached` here because it resets the flow loop state to ensure everything is computed from scratch before the type is permanently cached. The `getTypeOfExpression` cache is purely there to help performance, but there is no guarantee that the same type will be returned every time.